File: C++/Step19-Binary Search/1300.cpp
//
//  1300.cpp
//  Step19-Binary Search
//
//  Created by 김예빈 on 2019. 10. 4..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <algorithm>
using namespace std;

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    int n, k, result = 0, mid, left, right, cnt;
    
    cin >> n >> k;
    
    left = 1;
    right = k;
    
    while(left <= right){
        cnt = 0;
        mid = (left + right) / 2;
        
        for(int i = 1; i <= n; i++) {
            cnt += min(mid / i, n);
        }
        
        if(cnt < k)
            left = mid + 1;
        else {
            result = mid;
            right = mid - 1;
        }
    }
    
    cout << result;
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step20-Binary Search/1300.swift
//
//  1300.swift
//  Step20-Binary Search
//
//  Created by Yebin Kim on 2020/01/17.
//  Copyright © 2020 김예빈. All rights reserved.
//

let N = Int(readLine()!)!
let K = Int(readLine()!)!

let num = stride(from: 1, to: N + 1, by: 1)
var left = 1
var right = K
var answer = 0

while left <= right {
    var cnt = 0
    let mid = (left + right) / 2
    
    for e in num {
        cnt += min(N, mid / e)
    }
    
    if cnt < K {
        left = mid + 1
    } else {
        answer = mid
        right = answer - 1
    }
}

print(answer)
